src/backend/distributed/transaction/backend_data.c

```diff
@@ -0,0 +1,558 @@
+}
+
+
+/*
+ * get_all_active_transactions returns all the avaliable information about all
+ * the active backends.
+ */
+Datum
+get_all_active_transactions(PG_FUNCTION_ARGS)
+{
+	ReturnSetInfo *returnSetInfo = (ReturnSetInfo *) fcinfo->resultinfo;
+	TupleDesc tupleDescriptor = NULL;
+	Tuplestorestate *tupleStore = NULL;
+	MemoryContext perQueryContext = NULL;
+	MemoryContext oldContext = NULL;
+
+	List *activeBackendList = NULL;
+	ListCell *activeBackendCell = NULL;
+
+	const int attributeCount = 5;
+	Datum values[attributeCount];
+	bool isNulls[attributeCount];
+
+	CheckCitusVersion(ERROR);
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (returnSetInfo == NULL || !IsA(returnSetInfo, ReturnSetInfo))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context " \
+						"that cannot accept a set")));
+	}
+
+	if (!(returnSetInfo->allowedModes & SFRM_Materialize))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not " \
+						"allowed in this context")));
+	}
+
+	/* build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupleDescriptor) != TYPEFUNC_COMPOSITE)
+	{
+		elog(ERROR, "return type must be a row type");
+	}
+
+	perQueryContext = returnSetInfo->econtext->ecxt_per_query_memory;
+
+	oldContext = MemoryContextSwitchTo(perQueryContext);
+
+	tupleStore = tuplestore_begin_heap(true, false, work_mem);
+	returnSetInfo->returnMode = SFRM_Materialize;
+	returnSetInfo->setResult = tupleStore;
+	returnSetInfo->setDesc = tupleDescriptor;
+
+	MemoryContextSwitchTo(oldContext);
+
+	activeBackendList = GetAllActiveBackends();
+	activeBackendList = SortList(activeBackendList, CompareDistributedTransactionIds);
+
+	foreach(activeBackendCell, activeBackendList)
+	{
+		BackendData *backendData = (BackendData *) lfirst(activeBackendCell);
+		DistributedTransactionId *distributedTransctionId = &backendData->transactionId;
+
+		memset(values, 0, sizeof(values));
+		memset(isNulls, false, sizeof(isNulls));
+
+		values[0] = ObjectIdGetDatum(backendData->databaseId);
+		values[1] = Int32GetDatum(backendData->processId);
+		values[2] = Int32GetDatum(distributedTransctionId->initiatorNodeIdentifier);
+		values[3] = UInt64GetDatum(distributedTransctionId->transactionNumber);
+		values[4] = TimestampTzGetDatum(distributedTransctionId->timestamp);
+
+		tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+	}
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * GetAllActiveBackends returns all the active backends on the shared memory.
+ *
+ * Note that the function does not gurantee a global consistent result, the function
+ * only locks each active backend while copying it.
+ */
+static List *
+GetAllActiveBackends(void)
+{
+	int backendIndex = 0;
+	List *activeBackendList = NIL;
+
```

Comment:
Ok, so you suggest (i) inline `GetAllActiveBackends` into `get_all_active_transactions()`  (ii) protect via `lwlock`s.

Then, shall we use `tuplestore_putvalues` while the `lwlock` is held? Otherwise, the logic to fill `tupleStore` will become more complicated.